sdk/core/perf/src/base_test.cpp

```diff
@@ -96,12 +100,33 @@ class ProxyPolicy final : public HttpPolicy {
 
 namespace Azure { namespace Perf {
 
+  void BaseTest::ConfigureInsecureOptions(Azure::Core::_internal::ClientOptions* clientOptions)
```

Comment:
This method name seems wrong. It implies that this is a method which configures the insecure options, but it doesn't actually do that.

Instead, it seems to disable TLS certificate checks, so maybe that's a better method name?

Reply:
Within the `Performance Framework domain`, disabling TLS certificate check is known as `run insecure`.  

All languages use the input parameter `--insecure` to refer to this. So I am here following Lord @mikeharder  :)

You make a good point, @LarryOsterman. The documentation for the argument is: `Allow untrusted SSL certs. Default to false.`  So, that should help consumers to better understand what's the `insecure` option

Reply:
Then something like `ConfigurePipelineForInsecureOperations` makes more sense. You're not configuring the insecure options - there are no insecure options, you're configuring the pipeline to enable insecure operations.

Reply:
Pipeline is a concept for azure-core.
Within azure-perf, the concept is known as `insecure`.  
Developers from any other sdk language should be able to see the same option/concept across languages.  
This is not exposed to customers. This is an internal library to support performance testing for packages.
See: https://github.com/Azure/azure-sdk-for-net/blob/main/common/Perf/Azure.Test.Perf/PerfOptions.cs#L15
and
https://github.com/Azure/azure-sdk-for-net/blob/f4913258f0fd7db99d40b0999fcc56f035b673ac/common/Perf/Azure.Test.Perf/BatchPerfTest.cs#L37

Reply:
If there is a PerfOptions object, then why doesn't this method accept a PerfOptions then and set up a ClientOptions based on the status of the PerfOptions object? If PerfOptions derives from ClientOptions, that works as well - you accept a PerfOptions and then configure the ClientOptions portions of the PerfOptions to reflect the state of the PerfOptions fields.

Reply:
yeah. The design for the performance framework and test is not the best fit for C++, but I made it work :)
Let me explain why...

Performance tests are expected to be grouped by sdk-package within the same binary. So, a single binary is created per package.
The binary must take the name of the perf-test to be run as the first parameter.
After the first paramenter, the binary should be able to receive global-knwon arguments.  These are options that can be set to any perf-test, no matter that test or package we run. These options are the _insecure, parallel, iterations, warmup, etc_
After the global-knwon arguments, the binary can receive test-specific arguments!!!  A test can define a set of options that can be set only for that test.  For other languages, this is trivial to do using `reflection` to create the final test options based on the test definition and the global-options.

Well, for C++ there is no `reflections`.  So I use callbacks to define the tests, and the binary holds a map of tests, where each test is a callback which defines how to create the test and a callback which defines its specific options.  So, based on the input arguments, we can choose a test from the map and invoke the callbacks to create the test and know the options.
So, as you see, it will all be resolved during run-time :( .   That's basically the reason why we can't have a compile-time structure to define all the options XD.

Reply:
Ok, I've spent a bunch of time staring at the perf test code, and I think I now understand my core issue.

Why is `ConfigureInsecureOptions` a separate function rather than just a part of the existing `ConfigureCoreClientOptions`?  Based on my reading of the code, it seems like this is exactly the kind of thing which is supposed to be handled inside ConfigureCoreClientOptions.

Reply:
> Ok, I've spent a bunch of time staring at the perf test code

❤️ 

> Why is ConfigureInsecureOptions a separate function rather than just a part of the existing ConfigureCoreClientOptions?

That's to re-use the code.  A performance tests creates a raw-pipeline to send a request to the proxy-test server. When working with the test-proxy (this is also an azure-perf concept, as it not a general purpose proxy), the perf-test sends a request to ask the test-proxy to start recording, then another request to start playback and one more at the end to stop playback.

The function `ConfigureClientOptions()` adds the proxy-policy to the client options and then calls `ConfigureInsecureOptions()`.  When creating a raw-pipeline to sed a request within the perf-test, we don't need to add the proxy-policy again, so we only call `ConfigureInsecureOptions()` directly.

> Why isn't ConfigureCoreClientOptions called at all?

There are 2 methods exposed to customers that are creating a perf-test.  They must call either one of the other from the perf-test to create an SDK client. The first method is `ConfigureClientOptions(Azure::Core::ClientOptions options&)`. The method is useful when the customer needs to set some options before asking the perf-fw to set any required options. Like:

```cpp
    SomeClientOptions options
    options.Foo = Bar;
    // Then we call Configure
    ConfigureClientOptons(options);
    // then create the client
    SomeClient client(url, credentials, options);
```

The second method is for when the user is not interested in setting any option.  It is the `InitClientOptions()` and it can become above code to:
```cpp
// Create the client with options
SomeClient client(url, credentials, InitClientOptions<SomeClientOptions>());
```
The method handles creating the options, then calling Configure and returns the options as `rvalue` to the client constructor to take a copy of it :)

The reason the first method is not called, is because only the `InitClientOptions()` is currently used by test cases.

Reply:
Ok, So the plan is to add `ConfigurePortOptions`, `ConfigureLatencyOptions`, etc. methods  in the future?

Reply:
> Ok, So the plan is to add ConfigurePortOptions, ConfigureLatencyOptions, etc. methods in the future?

Consider it as implementation details.  There's not a contract to make it in an specific way.

Reply:
This entire pull request code review has been about the implementation details. Because to me, the implementation details matter a lot, especially when it comes to discoverability of functionality.

`ConfigureInsecureClientOptions` as a separate API makes sense *if* the design pattern for this class is to have a `ConfigureXxxClientOptions` where Xxx is some element of the standard perf test options. In other words, the `ConfigureXxxClientOptions` method exists to make whatever configuration changes are required to implement perf test option Xxx.

One other thought that would make things more understandable to me: If the method was named `ConfigureInsecureClientOptions`, it would have made understanding its function a lot easier - Based on the name of the method, I was looking for a set of InsecureOptions which were being set and tried to find where InsecureOptions was defined in this pull request. Naming the method using a pattern like "ConfigureXxxClientOptions" makes it clearer that what is being configured is a ClientOptions object. 

In this model, you'd have ConfigureClientOptions - this is the public facing method.
Within ConfigureClientOptions, you call a number of `Configure<something>ClientOptions` methods, each of which does configuration for a different set of ClientOptions (ConfigureCoreClientOptions, ConfigureInsecureClientOptions, etc).